test/select.jl

```diff
@@ -1306,4 +1327,246 @@ end
+                  DataFrame(a_function=[[1, 2], [2, 3]], x1=[1, 2], x2=[2, 3],
+                            p=[1, 2], q=[2, 3], s=[1, 2], t=[2, 3], k=[1, 2], l=[2, 3],
+                            z=[(s=1, t=2), (s=2, t=3)])
+            @test transform(df, :a => ByRow(x -> [x,x+1]),
+                            :a => ByRow(x -> [x, x+1]) => AsTable,
+                            :a => ByRow(x -> [x, x+1]) => [:p, :q],
+                            :a => ByRow(x -> (s=x, t=x+1)) => AsTable,
+                            :a => (x -> (k=x, l=x.+1)) => AsTable,
+                            :a => ByRow(x -> (s=x, t=x+1)) => :z) ==
+                  [df DataFrame(a_function=[[1, 2], [2, 3]], x1=[1, 2], x2=[2, 3],
+                                p=[1, 2], q=[2, 3], s=[1, 2], t=[2, 3], k=[1, 2], l=[2, 3],
+                                z=[(s=1, t=2), (s=2, t=3)])]
+            @test_throws ArgumentError select(df, :a => (x -> [(a=1,b=2), (a=1, b=2, c=3)]) => AsTable)
+            @test_throws ArgumentError select(df, :a => (x -> [(a=1,b=2), (a=1, c=3)]) => AsTable)
+            @test_throws ArgumentError combine(df, :a => (x -> (a=1,b=2)) => :x)
+        end
+    end
+
+    @testset "check correctness of duplicate column names" begin
+        for df in (DataFrame(a=1:2, b=3:4, c=5:6), view(DataFrame(a=1:3, b=3:5, c=5:7, d=11:13), 1:2, 1:3))
+            @test select(df, :b, :) == DataFrame(b=3:4, a=1:2, c=5:6)
+            @test select(df, :b => :c, :) == DataFrame(c=3:4, a=1:2, b=3:4)
+            @test_throws ArgumentError select(df, :b => [:c, :d], :)
+            @test_throws ArgumentError select(df, :a, :a => x -> (a=[1,2], b=[3,4]))
+            @test_throws ArgumentError select(df, :a, :a => (x -> (a=[1,2], b=[3,4])) => AsTable)
+            @test select(df, [:b, :a], :a => (x -> (a=[11,12], b=[13,14])) => AsTable, :) ==
+                  DataFrame(b=[13, 14], a=[11, 12], c=[5, 6])
+            @test select(df, [:b, :a], :a => (x -> (a=[11,12], b=[13,14])) => [:b, :a], :) ==
+                  DataFrame(b=[11, 12], a=[13, 14], c=[5, 6])
+        end
+    end
+end
+
+@testset "empty ByRow" begin
+    df = DataFrame(a=1:3)
+
+    @test select(df, [] => ByRow(() -> 1)) == DataFrame("function" => [1, 1, 1])
+    @test combine(df, [] => ByRow(() -> 1)) == DataFrame("function" => [1, 1, 1])
+    @test transform(df, [] => ByRow(() -> 1)) == DataFrame("a" => 1:3, "function" => [1, 1, 1])
+
+    for df in (DataFrame(), DataFrame(a=[]))
+        @test select(df, [] => ByRow(() -> 1)) == DataFrame("function" => [])
+        @test combine(df, [] => ByRow(() -> 1)) == DataFrame("function" => [])
+        if ncol(df) == 0
+            @test transform(df, [] => ByRow(() -> 1)) == DataFrame("function" => [])
+        else
+            @test transform(df, [] => ByRow(() -> 1)) == DataFrame("a" => [], "function" => [])
+        end
+        @test eltype(select(df, [] => ByRow(() -> 1)).function) == Int
```

Comment:
Maybe define a helper and use it everywhere to check both the contents and the type in a single line everywhere? IIRC we have something like this somewhere already.

Reply:
OK

Reply:
OK - will add (we have it for split-apply-combine)